export type Slotbar = {
	UUID: string,
	UnitName: string,
	Traits: { [number]: string },
	Relic: string,
	Skin: string,
	Level: number,
	XP: number,
	AbilityTree: { [string]: any },
	StatsPotential: { [string]: any },
	StatsTraining: { [string]: any },
	Locked: boolean,
	Shiny: boolean,
}

export type SlotbarData = {
	Slot1: Slotbar,
	Slot2: Slotbar,
	Slot3: Slotbar,
	Slot4: Slotbar,
	Slot5: Slotbar,
	Slot6: Slotbar,
}

local function createSlotbarData(): SlotbarData
	return {
		Slot1 = {
			UUID = "",
			UnitName = "",
			Traits = {
				[1] = "",
				[2] = "",
				[3] = "",
			},
			Relic = "",
			Skin = "",
			Level = 1,
			XP = 0,
			AbilityTree = {},
			StatsPotential = {},
			StatsTraining = {},
			Locked = false,
			Shiny = false,
		},
		Slot2 = {
			UUID = "",
			UnitName = "",
			Traits = {
				[1] = "",
				[2] = "",
				[3] = "",
			},
			Relic = "",
			Skin = "",
			Level = 1,
			XP = 0,
			AbilityTree = {},
			StatsPotential = {},
			StatsTraining = {},
			Locked = false,
			Shiny = false,
		},
		Slot3 = {
			UUID = "",
			UnitName = "",
			Traits = {
				[1] = "",
				[2] = "",
				[3] = "",
			},
			Relic = "",
			Skin = "",
			Level = 1,
			XP = 0,
			AbilityTree = {},
			StatsPotential = {},
			StatsTraining = {},
			Locked = false,
			Shiny = false,
		},
		Slot4 = {
			UUID = "",
			UnitName = "",
			Traits = {
				[1] = "",
				[2] = "",
				[3] = "",
			},
			Relic = "",
			Skin = "",
			Level = 1,
			XP = 0,
			AbilityTree = {},
			StatsPotential = {},
			StatsTraining = {},
			Locked = false,
			Shiny = false,
		},
		Slot5 = {
			UUID = "",
			UnitName = "",
			Traits = {
				[1] = "",
				[2] = "",
				[3] = "",
			},
			Relic = "",
			Skin = "",
			Level = 1,
			XP = 0,
			AbilityTree = {},
			StatsPotential = {},
			StatsTraining = {},
			Locked = false,
			Shiny = false,
		},
		Slot6 = {
			UUID = "",
			UnitName = "",
			Traits = {
				[1] = "",
				[2] = "",
				[3] = "",
			},
			Relic = "",
			Skin = "",
			Level = 1,
			XP = 0,
			AbilityTree = {},
			StatsPotential = {},
			StatsTraining = {},
			Locked = false,
			Shiny = false,
		},
	}
end

return createSlotbarData
